src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs

```diff
@@ -24,61 +24,110 @@ internal static class HPackEncoder
         {
-            Debug.Assert(value != null);
-            return EncodeHeader(name, value, null, null, buffer, out length);
-        }
+            int i = 0;
+            length = 0;
+            int valueLength = (value == null ? 0 : value.Length);
 
-        public static bool EncodeHeader(Headers.HeaderDescriptor header, string value, Span<byte> buffer, out int length)
-        {
-            Debug.Assert(value != null);
-            return EncodeHeader(header.Name, value, null, null, buffer, out length);
-        }
+            // We need at least \0 and twice length plus one octet string
+            if (buffer.Length < 3 + name.Length + (value == null ? 0 : value.Length))
+            {
+                return false;
+            }
 
-        public static bool EncodeHeader(Headers.HeaderDescriptor header, string[] values, string separator, Span<byte> buffer, out int length)
-        {
-            Debug.Assert(values != null);
-            Debug.Assert(values.Length > 0);
-            return EncodeHeader(header.Name, values, separator, buffer, out length);
-        }
+            buffer[i++] = 0;
+            if (!EncodeHeaderName(name, buffer.Slice(i), out int encodedLength))
+            {
+                return false;
+            }
 
-        public static bool EncodeHeader(string name, string[] values, string separator, Span<byte> buffer, out int length)
-        {
-            Debug.Assert(values != null);
-            Debug.Assert(values.Length > 0);
-            if (values.Length == 1)
+            i += encodedLength;
+
+            if (!IntegerEncoder.Encode(valueLength, 7, buffer.Slice(i), out encodedLength))
             {
-                return EncodeHeader(name, values[0], null, null, buffer, out length);
+                return false;
             }
 
-            // When we have more values, separator must be provided.
-            Debug.Assert(separator != null);
-            return EncodeHeader(name, null, values, separator, buffer, out length);
+            i += encodedLength;
+
+            if (valueLength > 0)
+            {
+                if (!EncodeStringPart(value, buffer.Slice(i), out encodedLength))
+                {
+                    return false;
+                }
+
+                i += encodedLength;
+            }
+
+            length = i;
+            return true;
         }
 
-        private static bool EncodeHeader(string name, string value, string[] values, string separator, Span<byte> buffer, out int length)
+        public static bool EncodeHeader(string name, string[] values, string separator, Span<byte> buffer, out int length)
         {
             int i = 0;
+            int valueLength = 0;
             length = 0;
 
-            // We need at least \0 and twice length plus one octet string
-            if (buffer.Length < 5)
+            if (values == null || values.Length == 0)
```

Comment:
Can values actually be null here?

Reply:
I was not sure but since we also check for empty array it seems like it would be cheap and safe to keep it in.